File: jetbrains/src/main/kotlin/org/example/smelly/BrainMethodExample.kt
package org.example.smelly

/**
 * Brain Method - A method that does too much and has high complexity.
 * CodeScene flags methods with high cyclomatic complexity and excessive lines.
 */
class BrainMethodExample {

    fun processOrder(
        orderId: String,
        customerId: String,
        items: List<String>,
        paymentMethod: String,
        shippingAddress: String,
        isExpress: Boolean
    ) {
        if (orderId.isBlank()) {
            System.err.println("Invalid order ID")
            return
        }
        if (customerId.isBlank()) {
            System.err.println("Invalid customer ID")
            return
        }

        val customer = when {
            customerId.startsWith("PREM") -> lookupPremiumCustomer(customerId)
                ?: run { System.err.println("Premium customer not found"); return }
            customerId.startsWith("REG") -> lookupRegularCustomer(customerId)
                ?: run { System.err.println("Regular customer not found"); return }
            else -> lookupGuestCustomer(customerId)
        }

        var subtotal = 0.0
        for (item in items) {
            subtotal += when {
                item.startsWith("BOOK") -> 15.99
                item.startsWith("DVD") -> 19.99
                item.startsWith("GAME") -> 59.99
                item.startsWith("TOY") -> 29.99
                else -> 9.99
            }
        }

        val discount = when {
            customer["type"] == "PREMIUM" -> when {
                subtotal > 100 -> subtotal * 0.20
                subtotal > 50 -> subtotal * 0.15
                else -> subtotal * 0.10
            }
            customer["type"] == "REGULAR" -> when {
                subtotal > 100 -> subtotal * 0.10
                subtotal > 50 -> subtotal * 0.05
                else -> 0.0
            }
            else -> 0.0
        }

        val shipping = when {
            isExpress -> when {
                "NY" in shippingAddress -> 15.99
                "CA" in shippingAddress -> 19.99
                else -> 24.99
            }
            else -> when {
                "NY" in shippingAddress -> 5.99
                "CA" in shippingAddress -> 7.99
                else -> 9.99
            }
        }

        val tax = when {
            "NY" in shippingAddress -> subtotal * 0.08875
            "CA" in shippingAddress -> subtotal * 0.0725
            "TX" in shippingAddress -> subtotal * 0.0625
            else -> subtotal * 0.06
        }

        val total = subtotal - discount + tax + shipping

        val paymentSuccess = when (paymentMethod) {
            "CREDIT_CARD" -> if (total < 10000) processCreditCard(customerId, total) else {
                System.err.println("Amount too high for credit card"); return
            }
            "PAYPAL" -> processPayPal(customerId, total)
            "BITCOIN" -> if (total > 100) processBitcoin(customerId, total) else {
                System.err.println("Bitcoin payment too small"); return
            }
            else -> { System.err.println("Unknown payment method"); return }
        }

        if (!paymentSuccess) {
            System.err.println("Payment failed")
            sendPaymentFailureEmail(customerId)
            logPaymentFailure(orderId, customerId, total)
            return
        }

        items.forEach { updateInventory(it, -1) }
        sendOrderConfirmationEmail(customerId, orderId, total)
        if (isExpress) sendExpressShippingEmail(customerId, orderId)
        logOrderCreated(orderId, customerId, total)
        logPaymentProcessed(orderId, paymentMethod, total)
        logShippingScheduled(orderId, shippingAddress, isExpress)
        updateCustomerOrderCount(customerId)
        updateCustomerLifetimeValue(customerId, total)
        println("Order processed successfully: $orderId")
    }

    private fun lookupPremiumCustomer(id: String): Map<String, Any?>? = emptyMap()
    private fun lookupRegularCustomer(id: String): Map<String, Any?>? = emptyMap()
    private fun lookupGuestCustomer(id: String): Map<String, Any?> = emptyMap()
    private fun processCreditCard(customerId: String, amount: Double): Boolean = true
    private fun processPayPal(customerId: String, amount: Double): Boolean = true
    private fun processBitcoin(customerId: String, amount: Double): Boolean = true
    private fun sendPaymentFailureEmail(customerId: String) {}
    private fun sendOrderConfirmationEmail(customerId: String, orderId: String, total: Double) {}
    private fun sendExpressShippingEmail(customerId: String, orderId: String) {}
    private fun logPaymentFailure(orderId: String, customerId: String, total: Double) {}
    private fun logOrderCreated(orderId: String, customerId: String, total: Double) {}
    private fun logPaymentProcessed(orderId: String, method: String, total: Double) {}
    private fun logShippingScheduled(orderId: String, address: String, express: Boolean) {}
    private fun updateInventory(item: String, change: Int) {}
    private fun updateCustomerOrderCount(customerId: String) {}
    private fun updateCustomerLifetimeValue(customerId: String, amount: Double) {}
}

File: vsc/src/smelly/BrainMethodExample.js
/**
 * Brain Method - A method that does too much and has high complexity.
 * CodeScene flags methods with high cyclomatic complexity and excessive lines.
 */

export class BrainMethodExample {
  processOrder(
    orderId,
    customerId,
    items,
    paymentMethod,
    shippingAddress,
    isExpress
  ) {
    if (!orderId?.trim()) {
      console.error("Invalid order ID");
      return;
    }
    if (!customerId?.trim()) {
      console.error("Invalid customer ID");
      return;
    }

    let customer;
    if (customerId.startsWith("PREM")) {
      customer = this.lookupPremiumCustomer(customerId);
      if (!customer) {
        console.error("Premium customer not found");
        return;
      }
    } else if (customerId.startsWith("REG")) {
      customer = this.lookupRegularCustomer(customerId);
      if (!customer) {
        console.error("Regular customer not found");
        return;
      }
    } else {
      customer = this.lookupGuestCustomer(customerId);
    }

    let subtotal = 0;
    for (const item of items) {
      if (item.startsWith("BOOK")) subtotal += 15.99;
      else if (item.startsWith("DVD")) subtotal += 19.99;
      else if (item.startsWith("GAME")) subtotal += 59.99;
      else if (item.startsWith("TOY")) subtotal += 29.99;
      else subtotal += 9.99;
    }

    let discount = 0;
    if (customer.type === "PREMIUM") {
      if (subtotal > 100) discount = subtotal * 0.2;
      else if (subtotal > 50) discount = subtotal * 0.15;
      else discount = subtotal * 0.1;
    } else if (customer.type === "REGULAR") {
      if (subtotal > 100) discount = subtotal * 0.1;
      else if (subtotal > 50) discount = subtotal * 0.05;
    }

    let shipping = 0;
    if (isExpress) {
      if (shippingAddress.includes("NY")) shipping = 15.99;
      else if (shippingAddress.includes("CA")) shipping = 19.99;
      else shipping = 24.99;
    } else {
      if (shippingAddress.includes("NY")) shipping = 5.99;
      else if (shippingAddress.includes("CA")) shipping = 7.99;
      else shipping = 9.99;
    }

    let tax = 0;
    if (shippingAddress.includes("NY")) tax = subtotal * 0.08875;
    else if (shippingAddress.includes("CA")) tax = subtotal * 0.0725;
    else if (shippingAddress.includes("TX")) tax = subtotal * 0.0625;
    else tax = subtotal * 0.06;

    const total = subtotal - discount + tax + shipping;

    let paymentSuccess = false;
    if (paymentMethod === "CREDIT_CARD") {
      if (total < 10000)
        paymentSuccess = this.processCreditCard(customerId, total);
      else {
        console.error("Amount too high for credit card");
        return;
      }
    } else if (paymentMethod === "PAYPAL") {
      paymentSuccess = this.processPayPal(customerId, total);
    } else if (paymentMethod === "BITCOIN") {
      if (total > 100) paymentSuccess = this.processBitcoin(customerId, total);
      else {
        console.error("Bitcoin payment too small");
        return;
      }
    } else {
      console.error("Unknown payment method");
      return;
    }

    if (!paymentSuccess) {
      console.error("Payment failed");
      this.sendPaymentFailureEmail(customerId);
      this.logPaymentFailure(orderId, customerId, total);
      return;
    }

    items.forEach((item) => this.updateInventory(item, -1));
    this.sendOrderConfirmationEmail(customerId, orderId, total);
    if (isExpress) this.sendExpressShippingEmail(customerId, orderId);
    this.logOrderCreated(orderId, customerId, total);
    this.logPaymentProcessed(orderId, paymentMethod, total);
    this.logShippingScheduled(orderId, shippingAddress, isExpress);
    this.updateCustomerOrderCount(customerId);
    this.updateCustomerLifetimeValue(customerId, total);
    console.log(`Order processed successfully: ${orderId}`);
  }

  lookupPremiumCustomer(id) {
    return { type: "PREMIUM" };
  }
  lookupRegularCustomer(id) {
    return { type: "REGULAR" };
  }
  lookupGuestCustomer(id) {
    return { type: "GUEST" };
  }
  processCreditCard() {
    return true;
  }
  processPayPal() {
    return true;
  }
  processBitcoin() {
    return true;
  }
  sendPaymentFailureEmail() {}
  sendOrderConfirmationEmail() {}
  sendExpressShippingEmail() {}
  logPaymentFailure() {}
  logOrderCreated() {}
  logPaymentProcessed() {}
  logShippingScheduled() {}
  updateInventory() {}
  updateCustomerOrderCount() {}
  updateCustomerLifetimeValue() {}
}
